### DISCUSSION TITLE
Developing extensions, some doubts

### DISCUSSION THREAD

**Discussion by ProvoK** [emad-elsaid/xlog#34] [General]
Hi there 👋 

I'm trying to use xlog for my digital gardening, some feature I really need are missing and I'm looking into writing custom extensions to fill the gap.
I would like to start thanking you for your work 🙇 

Note: I come from Logseq, which is pretty much fully featured, but misses the "webapp" approach which is essential to me! I don't want to sync my files for every device (and all the sync issues that comes with it. Your approach is much better imho, and it's somethin very requested in the logseq community.

# List some of missing features (for my use-cases)

Please note I'm not requesting anything but pointers on how to personally develop my extensions 🙏 

## Journaling
Each day I would like to log my activities, which could be further detailed in ad hoc pages. 
It's probably doable by registering a new command that opens a new page (`journal-{2006-01-02}` or similar) and hopefully uses a custom template for bootstrapping the page content.

While I was able to register the command, is unclear to me how I could open a new/existing page in editing with custom content.

## Dynamic content via queries

I was used to query content and embed it in page via https://docs.logseq.com/#/page/queries.
I believe this could be doable by implementing `/{something}` which dynamically renders traverse all the pages AST and produce <some> content.

Is there a better approach for finding all the TODOs other than using `EachPage` routine and traversing all the pages AST?
Also, how can I register an handler for `/{something}`?  I wasn't able to find any `Register*` function to do that.

## Tasks
Usually, you are meant to do something within a certain deadline, and I would like to develop something like https://docs.logseq.com/#/page/tasks 

As far as I can see, there is no possibility in xlog to register generic "actions" that results in producing a dynamic content.
Eg: `/today` that renders `2023-08-24` (autocomplete can't do this, since suggestion are evaluated at page load)

Also, how to implement "tasks" such as logseq without extending Markdown syntax?

---

These are just the top 3 things, there could have been more but for sake of sanity I won't write all them down 😄 
I've also some smaller feedbacks on usability/UX, where could I share them?

**Comment by emad-elsaid** [comment]
thanks for the lengthy description here are some ideas that may help: 
* to make the command open a URL you'll need to define a `Link` method for your command https://pkg.go.dev/github.com/emad-elsaid/xlog#Command
* for Journaling you actually don't need to extend xlog at all, most browser allow you to add a bookmark that the value is `javascript:` instead of a URL, you can make a bookmark that opens a URL that has the current date.
* Queries are complicated so I won't write my ideas about it now. I'll think about it for a while
* To register a Get or Post handler you can use `Get` `Post` `Match` ...etc https://pkg.go.dev/github.com/emad-elsaid/xlog#Get 
* For tasks. there is already an extension developed for tasks it will even show you count of tasks that are done / all tasks beside the page link. https://xlog.emadelsaid.com/Features%20Test/#task-lists https://github.com/emad-elsaid/xlog/tree/master/extensions/todo

**Comment by ProvoK** [comment]
Thanks for the reply @emad-elsaid !



> For tasks. there is already an extension developed for tasks it will even show you count of tasks that are done / all tasks beside the page link. https://xlog.emadelsaid.com/Features%20Test/#task-lists https://github.com/emad-elsaid/xlog/tree/master/extensions/todo

Yup I saw that, but current todos have no possibility to attach a "deadline" to them

**Reply by emad-elsaid** [reply]
it's true tasks/todos don't have a deadline attached to it. but also xlog recognise [date formats](https://xlog.emadelsaid.com/Features%20Test/#:~:text=%F0%9F%94%A7-,Detect%20date,-DD%2DMM%2DYYYY). and every date has dedicated page that lists all pages that include [this date](https://xlog.emadelsaid.com/+/date/1-1-2023/) so maybe we can improve that? 
maybe we can also add a shortcode that list todos for specific date or relative date, which then query pages and lists todo items that has a date struct inside it that matches the required date? something similar to hashtag shortcode.

**Reply by ProvoK** [reply]
I'll try to think of a nice way to manage this, not on top of my head atm

**Comment by ProvoK** [comment]
> for Journaling you actually don't need to extend xlog at all, most browser allow you to add a bookmark that the value is javascript: instead of a URL, you can make a bookmark that opens a URL that has the current date.

Actually, I need to open a new page with some default content and structure, so simply editing a new page won't do the trick Im afraid.
I tried implementing a `Get` for this, but it's not possible to register routes with wildcards or similar, since xlog uses default golang http handler func.

**Reply by emad-elsaid** [reply]
xlog mux matches with regex. so you can have a wildcard for sure. this is how many pages are implemented in extensions. checkout the hashtags [extension for example](https://github.com/emad-elsaid/xlog/blob/master/extensions/hashtags/hashtags.go#L27) it matches a segment with `{tag}` and the edit pages [here](https://github.com/emad-elsaid/xlog/blob/master/handlers.go#L15) matches a path assign it to variable `page` and has regex after it to allow any character even `/`. so it matches the rest of the path in this case. instead of `.+` you can put your own regex or leave it empty to match any character except for `/` . implementation [here](https://github.com/emad-elsaid/xlog/blob/master/server.go#L129)

**Reply by ProvoK** [reply]
thank you very much! I've totally missed that 😄

**Reply by ProvoK** [reply]
did a quick test, works as a charm 💯

**Reply by emad-elsaid** [reply]
let me know when you have a POC, try to have a new repo for the extension. it should help if you're going to include it in the official distribution of Xlog in the future

**Reply by ProvoK** [reply]
sure! I'm currently reading "building a second brain" from Tiago Forte, and thinking of developing a single extension to, in some way, be able to follow the [PARA method](https://fortelabs.com/blog/para/).
I believe a proper use of hashtags, page templates and ad hoc search queries could do the trick, at least for starting off.

I'll keep you posted, unfortunately I don't have too much time atm for coding
